File: server/src/main/java/ch/zxseitz/tbsg/server/websocket/Protector.java
package ch.zxseitz.tbsg.server.websocket;

import java.util.concurrent.locks.Lock;
import java.util.concurrent.locks.ReentrantLock;

public class Protector<T extends Comparable<T>> implements IProtectable<T> {
    private final Lock lock;
    private final T t;

    public Protector(T t) {
        this.lock = new ReentrantLock();
        this.t = t;
    }

    public T get() {
        return t;
    }

    @Override
    public void lock() {
        lock.lock();
    }

    @Override
    public void unlock() {
        lock.unlock();
    }

    @Override
    public int compareTo(T o) {
        return t.compareTo(o);
    }
}

File: server/src/main/java/ch/zxseitz/tbsg/server/websocket/Protector.kt
package ch.zxseitz.tbsg.server.websocket;

import java.util.concurrent.locks.Lock;
import java.util.concurrent.locks.ReentrantLock;

class Protector<T: Comparable<T>>(val t: T): IProtectable<T> {
    private val lock: Lock = ReentrantLock()

    override fun lock() {
        lock.lock();
    }

    override fun unlock() {
        lock.unlock();
    }

    override fun compareTo(other: T): Int {
        return t.compareTo(other);
    }
}
